### ISSUE TITLE
`defaultProps` rule to be deprecated on function components

### ISSUE THREAD

**Issue by amypellegrini** [jsx-eslint/eslint-plugin-react#2396]
`defaultProps` on function components is getting deprecated by React (see: https://github.com/facebook/react/pull/16210)

For this rule to follow suit, it should only be triggered when missing from class components.

The official recommendation in future versions of React is to use ES6 default values for function components.

**Comment by ljharb** [comment]
That won't change by default; there can be an option to ignore function components, however.

**Comment by KevinGhadyani-minted** [comment]
Are you saying there's no way we could use object deconstruction assignment when requiring `defaultProps`?

**Comment by KevinGhadyani-minted** [comment]
Is there a way to check the deconstructed values for default props?

**Comment by ljharb** [comment]
@KevinGhadyani-minted I'm not sure what you're asking.

**Comment by KevinGhadyani-minted** [comment]
The change isn't to remove `defaultProps` from function components, but to use the official ES6 standard for defaulting props using object deconstruction and default assignment.

Where we once had:
```js
const TestComponent.defaultProps = {
  size: 'medium',
  type: 'primary',
}
```

We now have:
```js
const TestComponent = ({
  size = 'medium',
  type = 'primary',
}) => (
  // ...
)
```

So the goal isn't to remove `defaultProps`, but change how they're defined on function components.

I would expect the `defaultProps` rules to function the same regardless of if I use `defaultProps` or object deconstruction and default assignment.

**Comment by ljharb** [comment]
That's something that should be configurable via an option, as I said [above](https://github.com/yannickcr/eslint-plugin-react/issues/2396#issuecomment-526494760).

**Comment by KevinGhadyani-minted** [comment]
I think I'm confused then. In your comment, you said there _can_ be an option to ignore it for function components.

To me, this means there's development that would occur to disable it for function components, but this development has not occurred. It doesn't say there'd be an alternative for getting it to identify ES2015 default assignment instead.

**Comment by ljharb** [comment]
Yes, it means someone would have to do the development work to add an option.

It makes sense to me to modify `require-default-props` so that the default is "defaultProps" for everything, but that there's an option for function components to instead require "default arguments", or, to ignore function components.

**Comment by aaronmccall** [comment]
@ljharb, I'm considering working on the addition you suggested above, is there a good place to start beyond looking at the current rule implementation?

**Comment by ljharb** [comment]
@aaronmccall That, and its tests, are pretty much it :-)

**Comment by nknaveen007** [comment]
if you are using a state ,then use the default values like below,
const BlogPostForm=({onSubmit,initialValues=()=>{initialValues.title='',initialValues.content=''}})=>{      //default values set in a function

    const [title,setTitle]=useState(initialValues.title)
    const [content,setContent]=useState(initialValues.content)

**Comment by chbdetta** [comment]
Can we also have a way to enforce that function components can't have `defaultProps`?

**Comment by ljharb** [comment]
@chbdetta that's included in https://github.com/yannickcr/eslint-plugin-react/issues/2396#issuecomment-539688853

**Comment by flying-sheep** [comment]
My thoughts:

- Using defaultProps is fine when using JS as it can be used for checks
- Using destructured defaults is superior when using typescript, since typescript doesn’t need those checks and by using destructured defaults, the accepted prop types don’t differ from the object the function actually gets passed.

To demonstrate: Who wants to do this dance for every component? https://stackoverflow.com/a/69449759/247482

```ts
import * as React from "react";

interface IMyComponentRequiredProps {
  title: string;
}
interface IMyComponentOptionalProps {
  color: string;
  fontSize: number;
}
interface IMyComponentProps
  extends IMyComponentRequiredProps,
    IMyComponentOptionalProps {}

const defaultProps: IMyComponentOptionalProps = {
  color: "red",
  fontSize: 40,
};

const MyComponent = ({ title, color, fontSize }: IMyComponentProps) => {
  return <h1 style={{ color, fontSize }}>{title}</h1>;
};

MyComponent.defaultProps = defaultProps;

export default MyComponent;
```

The same with defaults:

```ts
import * as React from "react";

interface IMyComponentProps {
  title: string;
  color?: string;
  fontSize?: number;
}

const MyComponent = ({ title, color = "red", fontSize = 40 }: IMyComponentProps) => {
  return <h1 style={{ color, fontSize }}>{title}</h1>;
};

export default MyComponent;
```

My solution is therefore:

`eslintrc.yaml`
```yaml
rules:
  react/require-default-props:
  - error
  - forbidDefaultForRequired
overrides:
- files: ['*.ts', '*.tsx']
  rules:
    react/require-default-props: off # use default arguments in typescript
```

**Comment by ljharb** [comment]
Someone who wants maximal correctness would want to do that dance.

**Comment by flying-sheep** [comment]
The second variant is no less correct, it just doesn’t give some of the types names.

Regarding our comments being hidden: Is this not the right place for this discussion? Whatever is the case, I very much think that my `eslintrc.yaml` might be helpful to people so in hopes this comment won’t be hidden without explanation I’ll reproduce it here:

```yaml
rules:
  react/require-default-props:
  - error
  - forbidDefaultForRequired
overrides:
- files: ['*.ts', '*.tsx']
  rules:
    react/require-default-props: off # use default arguments in typescript
```

**Comment by ljharb** [comment]
@flying-sheep discussion about just disabling the rule, even with overrides, won’t likely help anyone; people already know they can do that.

Either way, defaultProps should absolutely be used in TS still, and the type info does *not* contain the same information - because 100% of it is stripped out at runtime, which is when the defaultProps information can be needed.

**Comment by amypellegrini** [comment]
> @flying-sheep discussion about just disabling the rule, even with overrides, won’t likely help anyone; people already know they can do that.
> 
> Either way, defaultProps should absolutely be used in TS still, and the type info does _not_ contain the same information - because 100% of it is stripped out at runtime, which is when the defaultProps information can be needed.

Would that very depending on the compilation target? I would expect default values assigned in ES6 to still work at runtime (be that built by TS or not), right?

**Comment by ljharb** [comment]
@amypellegrini they work, but are not introspectable, so at runtime tooling can’t read information about what’s defaulted except with .defaultProps.

**Comment by flying-sheep** [comment]
Why treat components in a special way? If you use JS, your code base could benefit from runtime type checking for all kinds of functions, not just components.

Once you have fully migrated a code base to  TypeScript, all types are proven correct at compile time and runtime checking isn’t necessary.

**Comment by ljharb** [comment]
>  all types are proven correct at compile time and runtime checking isn’t necessary.

This is objectively false, since every system includes runtime user data (from the network, filesystem, user interaction, etc), as well as a majority of third-party code that was never typechecked, and this false belief is actually the worst damage caused by the use of a type system.

**Comment by ljharb** [comment]
Even if TS was a sound type system - which it is not - it would NEVER be safe to omit runtime checks or to believe that types are "proven" correct. TS, for example, thinks that this JS snippet which has thrown an exception since 2009 - `Object.create(null).toString()` - is perfectly type-safe, so it's incredibly easy to disprove your claim.

**Comment by flying-sheep** [comment]
Well, seems like none of my systems ever put user data through edge cases like the one you pointed out, but point taken!

**Comment by OwnerOfThisIsle** [comment]
@flying-sheep In your solution I would miss a linting rule to make sure the destructured defaults are really in place.

**Comment by flying-sheep** [comment]
No need, the type system takes over. E.g. if I have this:

```tsx
// required prop “x”
interface CompAProps { x: string }
const CompA = ({ x }: CompAProps) => <>{x}</>

// optional prop “y”
interface CompBProps { y?: string }
const CompB = ({ y }: CompBProps) => <CompA x={y} />
```

TypeScript will complain that `<CompA x={x} />` is invalid as the type of `y` is `string | undefined` and not `string`.
I can then choose to fix this however I want, e.g. like this:

```tsx
const CompB = ({ y = 'some default' }: CompBProps) => <CompA x={y} />
```

**Comment by uloco** [comment]
If this gets depreacted, is there still a way to set the defaultProps globally for example the react-native `<Text>` element?

**Comment by ljharb** [comment]
@uloco im not familiar with what you’re describing, but either way, that’s a question to ask react native, and only if concrete plans to depreciate it ever materialize.

**Comment by avin-kavish** [comment]
Any way to disable the error that react throws in development about this?

**Comment by pavinduLakshan** [comment]
> [#2396 (comment)](https://github.com/jsx-eslint/eslint-plugin-react/issues/2396#issuecomment-539688853) is the plan of record here.

Does anyone have an idea whether this improvement is available now? 

Thanks in Advance!

**Comment by ljharb** [comment]
@pavinduLakshan at the top, it says "fixed by #3249", and if you go there, you can see my comment https://github.com/jsx-eslint/eslint-plugin-react/pull/3249#issuecomment-1130645309 which says that it's in v7.30.0.

**Comment by pavinduLakshan** [comment]
> @pavinduLakshan at the top, it says "fixed by #3249", and if you go there, you can see my comment [#3249 (comment)](https://github.com/jsx-eslint/eslint-plugin-react/pull/3249#issuecomment-1130645309) which says that it's in v7.30.0.

Thanks. missed it.

**Comment by Spanking25** [comment]
> if you are using a state ,then use the default values like below, const BlogPostForm=({onSubmit,initialValues=()=>{initialValues.title='',initialValues.content=''}})=>{ //default values set in a function
> 
> ```
> const [title,setTitle]=useState(initialValues.title)
> const [content,setContent]=useState(initialValues.content)
> ```





https://reacter.space/questions/ways-to-specify-the-default-value-for-a-component
